design/kanister-kopia-integration.md

```diff
@@ -0,0 +1,54 @@
+# Integrating Kopia with Kanister
+
+This document proposes all the changes required within Kanister to better support the use of Kopia backup/restore tool.
+
+## Introducing Kopia
+
+[Kopia](https://kopia.io/) is a powerful, cross-platform tool for managing encrypted backups in the cloud.
+It provides fast and secure backups, using compression, data deduplication and client-side end-to-end encryption.
+It supports a variety of backup storage targets, including object stores, which allows users to choose the storage provider that better addresses their needs.
+It is a lock-free system that allows for concurrent multi-client operations including garbage collection.
+
+## Goal
+
+Kopia can focus on its core data export and transformation as part of various Kanister functions.
+While, Kanister continues to execute these functions from it's application-centric data protection workflows called Blueprints.
+
+##  Proposed Work
+
+### Kanister Functions 
+
+We already have a rich repository of Kanister functions present on path `pkg/function` that enable application-level data protection in various use cases.
+Few of these functions need to be re-worked to use Kopia. For starters, these functions are;
+
+1. BackupData
+2. BackupDataAll
+3. BackupDataStats
+4. CopyVolumeData
+5. DeleteData
+6. DeleteDataAll
+7. RestoreData
+8. RestoreDataAll
+
+Please note that the functions stated above will only be **refactored** to use Kopia underneath.
```

Comment:
Will we be able to use something besides Kopia or will this make Kopia the only mechanism supporting these functions?

Reply:
We can continue to maintain separate versions of these Kanister Functions. For ex; `v.0.0.0` of BackupData (or any other Function) will continue to follow restic workflow and a version like `v1.0.0-alpha` could be implemented using Kopia.

The reason I mentioned "refactor" is because in a previous discussion with @pavannd1, we thought one of the end goal here could be to completely replace the Restic based Kanister Functions with the one's that use Kopia. But if that's still unclear, we can decide on it at a later point since function versions are easily configurable by downstream users.

Reply:
Can we design the data-mover interface in such a way that in the future, if we want to switch to another tool it will be easier than refactoring all the functions? I understand this could be out of the current scope but just wondering.